src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala

```diff
@@ -0,0 +1,482 @@
+/**
+  * Copyright 2017 RiskSense, Inc.
+  * This file is part of ipaddr library.
+  *
+  * Ipaddr is free software licensed under the Apache License, Version 2.0 (the "License"); you
+  * may not use this file except in compliance with the License. You may obtain a copy of the
+  * License at
+  *
+  *         http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software distributed under the
+  * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+  * express or implied. See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package com.risksense.ipaddr
+
+import scala.annotation.tailrec
+import scala.collection.SortedSet
+import scala.collection.SortedSetOps
+import scala.collection.SpecificIterableFactory
+import scala.collection.StrictOptimizedSortedSetOps
+import scala.collection.mutable
+import scala.util.hashing.MurmurHash3
+
+/** Represents an unordered collection (set) of IpNetwork elements.
+  *
+  * @constructor creates a new IpSet.
+  * @param networkSeq a sorted sequence of [[IpNetwork]] objects.
+  */
+case class IpSet private[ipaddr] (networkSeq: IndexedSeq[IpNetwork]) // scalastyle:ignore
+  extends SortedSet[IpNetwork]
+    with SortedSetOps[IpNetwork, SortedSet, IpSet]
+    with StrictOptimizedSortedSetOps[IpNetwork, SortedSet, IpSet] {
+
+  /** Size of this IpSet.
+    *
+    * The cardinality of this IpSet (based on the number of individual IP addresses including those
+    * implicitly defined in networks).
+    */
+  lazy val volume: Long = networkSeq.map(_.size).sum[Long]
+
+  /** True if the members of the set form a contiguous IP address range (with no gaps),
+    * False otherwise.
+    */
+  lazy val isContiguous: Boolean = {
+    if (networkSeq.size > 1) {
+      val networkPairs = networkSeq.sliding(2)
+      networkPairs.forall { netSeq => (netSeq(0).last + 1).equals(netSeq(1).first) }
+    } else {
+      true
+    }
+  }
+
+  /** [[IpRange]] representation of this IpSet.
+    *
+    * IpRange equivalent to this IpSet only if all its members form a single contiguous sequence,
+    * otherwise generates a IpaddrException.
+    *
+    * @note This can be resource intensive if the address space is huge.
+    */
+  @throws(classOf[IpaddrException])
+  lazy val ipRange: IpRange = {
+    if (isContiguous) {
+      if (networkSeq.isEmpty) {
+        throw new IpaddrException("Cannot create IpRange from an empty IpSet.")
+      } else {
+        val ip1 = IpAddress(networkSeq.head.allHosts.head)
+        val ip2 = IpAddress(networkSeq.last.allHosts.last)
+        new IpRange(ip1, ip2)
+      }
+    } else {
+      throw new IpaddrException(
+        "The input IpRange does not represent a single contiguous sequence of addresses")
```

Comment:
I presume throwing is how the library currently behaves

Reply:
Yes